### ISSUE TITLE
Distillation: per-microbatch mean weights microbatches, not tokens, so unequal lengths still shift the gradient

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#932]
## What

#742 (merged `dd151374`) made distillation's accumulated gradient independent of `gradient_accumulation_steps`. It does **not** make it equal to the full-batch gradient when microbatches hold different numbers of tokens.

A per-microbatch mean weights each microbatch equally. The equivalent single full batch weights each **token** equally. With equal-length chunks the two coincide, which is why the merged tests — which use 8 rows split into 4 or 8 equal chunks — measure a clean `1.0` ratio. With unequal lengths they diverge.

## Why this is filed separately, and why it has no test

This residual is named in #720's own text. It **was** covered: `test_unequal_microbatch_lengths_still_match_within_the_mean_contract` existed in an earlier revision of #742.

I asked for that test to be deleted, along with four others, because they built a bare `torch.nn.Linear(4, 1)` and took a `compensate: bool` the test supplied itself — five of the six passed on unmodified `main` and would have passed with `src/soup_cli/` deleted. That judgement was right for the other four. For this one it was right about the *test* and wrong about the *consequence*: it was the only place in the repo stating the residual, and deleting it left the fact recorded nowhere.

So this issue exists because of a call I made in review, not because the contributor missed something. @dchaudhari7177's merged work is complete for what #720 asked.

## Acceptance criteria

- [ ] A test that measures the gradient with **unequal** microbatch lengths — e.g. 8 rows of lengths `[2, 2, 2, 2, 16, 16, 16, 16]` split into two chunks — and states the expected relationship to the full-batch gradient, whichever way the project decides.
- [ ] A decision, recorded in the code or the docs: either token-weight the loss so the full-batch equality holds for any chunking, **or** document that per-microbatch mean is the contract and unequal lengths shift the weighting. Both are defensible; leaving it unstated is not.
- [ ] If token-weighting is chosen, the equal-length case must be unchanged — the merged `1.0` ratio at GA=4 and GA=8 is the regression test for that.

## Notes for whoever takes this

Do **not** start by writing the fix. Start by writing the measurement: build the unequal-length batch, measure the ratio on current `main`, and put the number in the issue. If it turns out the divergence is negligible at realistic sequence lengths, the honest outcome is a documented contract rather than a code change, and that is a perfectly good result.

The existing test file shows the technique — it compiles the real `_DistillTrainer` class body out of `distill.py` via AST and runs `Trainer.training_step` on a 1-layer `LlamaForCausalLM`. Reuse it; do not write a look-alike subclass, which is exactly what made the deleted tests non-discriminating.

Not a good first issue: it needs a view on what the loss contract should be, not just a test.

**Comment by Amix29** [comment]
I’ll take this one. I’ll follow the requested order: first compile and exercise the real `_DistillTrainer` from current `origin/main` with unequal token counts, record the measured gradient relationship here, and only then choose and implement the loss contract. I’ll preserve the equal-length GA=4/8 regression and avoid a look-alike trainer.

Qualification will run on macOS without CUDA; the reproducer uses the tiny one-layer Llama setup already established by `tests/test_issue720_distill_grad_accum.py`.

**Comment by Amix29** [comment]
Measurement first, as requested — current `origin/main` at `e27e40c`, using the real `_DistillTrainer` class body compiled from `src/soup_cli/trainer/distill.py`, real `transformers.Trainer.training_step`, a 1-layer `LlamaForCausalLM`, CPU on Apple Silicon, `torch 2.12.1`, and the declared Transformers floor `5.16.1`. No CUDA path.

Eight rows had lengths `[2, 2, 2, 2, 16, 16, 16, 16]`. After causal shifting that is 64 trained targets total, split 4 vs 60 across the two microbatches. The single full batch and GA=2 runs used identical initialized weights and token ids.

```text
full gradient norm                 0.7767304778
GA=2 split gradient norm           1.3807157278
split / full norm ratio            1.7775995731
relative L2 error                  1.5322198868
cosine similarity                  0.5097218752
max absolute component difference  0.2587058246
```

So the divergence is not negligible and is not merely a scalar reporting difference: the gradient direction changes materially. I am taking the token-weighted contract. The implementation will make the trainer consume the full accumulation window's `num_items_in_batch` and scale each existing microbatch-mean loss by `local_shifted_trained_tokens / window_shifted_trained_tokens`. That preserves every existing loss branch while making accumulated gradients equal the full-batch token mean. The existing equal-length GA=4/8 test remains the regression control.

**Comment by Amix29** [comment]
Post-fix measurement from #943, same rows/seeds/model and the real Trainer window counter:

```text
sequence CE
  full norm          0.7767304778
  GA=2 norm          0.7767304778
  ratio              1.0000000000
  relative L2 error  7.31e-08
  max abs diff       1.86e-08

CE + teacher KL
  full norm          0.3881082535
  GA=2 norm          0.3881082535
  ratio              1.0000000000
  relative L2 error  8.09e-08
  max abs diff       9.31e-09
```

Both complete gradient vectors match within the test tolerance; the equal-length GA=4/8 controls remain at 1.0 as well.

**Comment by MakazhanAlpamys** [comment]
Fixed in #943 by **@Amix29**, merged as `4b93b4ab`.

Try: `soup train` with `task: distill` and `gradient_accumulation_steps > 1`.

`_DistillTrainer` weighted every microbatch equally across an accumulation window. Whenever the microbatches in a window carried different numbers of trained tokens — which is the normal case for variable-length sequences — the resulting gradient was **not** the full-batch gradient: short sequences over-weighted, long ones under-weighted.

It now consumes Transformers' full-window `num_items_in_batch` and scales each microbatch-mean loss by its share of trained tokens, which is the token-weighted full-batch contract rather than an approximation of it. This supersedes #742's fixed-per-microbatch compensation.

**Why this mattered more than the arithmetic suggests:** nothing about it is visible at runtime. The loss curve looks normal with the bug and without it, so a model trained this way is simply slightly wrong in a direction nobody would investigate.

Verified by mutation: replacing the weighting with `loss / gradient_accumulation_steps` mismatches **95.8%** and **94.4%** of gradient elements across the two comparison tests.
